src/profile-logic/track-initialization.js

```diff
@@ -0,0 +1,332 @@
+        globalTrack.mainThreadIndex = threadIndex;
+      }
+    } else {
+      // This is a non-main thread.
+      if (!globalTracksByPid.has(pid)) {
+        // This is a thread without a known main thread. Create a global process
+        // track for it, but don't add a main thread for it.
+        const globalTrack = {
+          type: 'process',
+          pid: pid,
+          mainThreadIndex: null,
+        };
+        globalTracks.push(globalTrack);
+        globalTracksByPid.set(pid, globalTrack);
+      }
+    }
+  }
+  return globalTracks;
+}
+
+export function initializeGlobalTrackOrder(
+  globalTracks: GlobalTrack[],
+  urlGlobalTrackOrder: TrackIndex[] | null
+): TrackIndex[] {
+  return urlGlobalTrackOrder !== null &&
+    _indexesAreValid(globalTracks.length, urlGlobalTrackOrder)
+    ? urlGlobalTrackOrder
+    : globalTracks.map((_, index) => index);
+}
+
+export function initializeSelectedThreadIndex(
+  selectedThreadIndex: ThreadIndex | null,
+  visibleThreadIndexes: ThreadIndex[],
+  profile: Profile
+): ThreadIndex | null {
+  if (
+    selectedThreadIndex !== null &&
+    visibleThreadIndexes.includes(selectedThreadIndex)
+  ) {
+    // This is a valid thread index to select.
+    return selectedThreadIndex;
+  }
+  // Select either the GeckoMain [tab] thread, or the first thread in the thread
+  // order.
+  const threadIndex = profile.threads.indexOf(
+    _findDefaultThread(
+      visibleThreadIndexes.map(threadIndex => profile.threads[threadIndex])
+    )
+  );
+  // The threadIndex being null should only happen in tests where profiles
+  // have no threads.
+  return threadIndex === -1 ? null : threadIndex;
+}
+
+export function initializeHiddenGlobalTracks(
+  globalTracks: GlobalTrack[],
+  threads: Thread[],
+  validTrackIndexes: TrackIndex[],
+  urlHiddenGlobalTracks: Set<TrackIndex> | null
+): Set<TrackIndex> {
+  const hiddenGlobalTracks = new Set();
+
+  if (urlHiddenGlobalTracks === null) {
+    // No hidden global tracks exist, generate the default Set.
+    for (const track of globalTracks) {
+      if (track.type === 'process' && track.mainThreadIndex !== null) {
+        const { mainThreadIndex } = track;
+        const thread = threads[mainThreadIndex];
+        if (_isThreadIdle(thread)) {
+          hiddenGlobalTracks.add(mainThreadIndex);
+        }
+      }
+    }
+    return hiddenGlobalTracks;
+  }
+
+  // Validate the global tracks to hide.
+  for (const trackIndex of urlHiddenGlobalTracks) {
+    if (validTrackIndexes.includes(trackIndex)) {
+      hiddenGlobalTracks.add(trackIndex);
+    }
+  }
+  return hiddenGlobalTracks;
+}
+
+export function getVisibleThreads(
+  globalTracks: GlobalTrack[],
+  hiddenGlobalTracks: Set<TrackIndex>,
+  localTracksByPid: Map<Pid, LocalTrack[]>,
+  hiddenTracksByPid: Map<Pid, Set<TrackIndex>>
+): ThreadIndex[] {
+  const visibleThreads = [];
+  for (const globalTrack of globalTracks) {
+    if (globalTrack.type === 'process') {
```

Comment:
could we have global tracks that aren't of type `process` ?

Reply:
Yes, screenshots will be a global track.